pkg/kubectl/cmd/create_role.go

```diff
@@ -0,0 +1,228 @@
+	cmd.Flags().StringSlice("resource", []string{}, "resource that the rule applies to")
+	cmd.Flags().StringSlice("resource-name", []string{}, "resource in the white list that the rule applies to")
+
+	return cmd
+}
+
+func (c *CreateRoleOptions) Complete(cmd *cobra.Command, args []string) error {
+	name, err := NameFromCommandArgs(cmd, args)
+	if err != nil {
+		return err
+	}
+	c.Name = name
+
+	// support specify multiple verbs together
+	// e.g. --verb=get,watch,list
+	verbs := cmdutil.GetFlagStringSlice(cmd, "verb")
+	for _, v := range verbs {
+		c.Verbs = mergeArrays(c.Verbs, strings.Split(v, ","))
+	}
+
+	// support specify multiple resources together
+	// e.g. --resource=pods,deployments.extensions
+	candidateResources := []string{}
+	resources := cmdutil.GetFlagStringSlice(cmd, "resource")
+	for _, r := range resources {
+		candidateResources = mergeArrays(candidateResources, strings.Split(r, ","))
+	}
+	for _, r := range candidateResources {
+		// support resource.group pattern
+		index := strings.Index(r, ".")
+		resourceName, apiGroup := "", ""
+
+		// No API Group specified, use "" as core API Group
+		if index == -1 {
+			resourceName = r
+		} else {
+			resourceName, apiGroup = r[0:index], r[index+1:]
+		}
+
+		if !arrayContains(c.APIGroups, apiGroup) {
+			c.APIGroups = append(c.APIGroups, apiGroup)
+		}
+		if !arrayContains(c.Resources, resourceName) {
+			c.Resources = append(c.Resources, resourceName)
+		}
+	}
+
+	// support specify multiple resource names together
+	// e.g. --resource-name=foo,boo
+	resourceNames := cmdutil.GetFlagStringSlice(cmd, "resource-name")
+	for _, n := range resourceNames {
+		c.ResourceNames = mergeArrays(c.ResourceNames, strings.Split(n, ","))
+	}
+
+	return nil
+}
+
+func (c *CreateRoleOptions) RunCreateRole(f cmdutil.Factory, cmdOut io.Writer, cmd *cobra.Command, args []string) error {
+	mapper, typer := f.Object()
+	dryRun, outputFormat := cmdutil.GetDryRunFlag(cmd), cmdutil.GetFlagString(cmd, "output")
+
+	var generator kubectl.StructuredGenerator
+	switch generatorName := cmdutil.GetFlagString(cmd, "generator"); generatorName {
+	case cmdutil.RoleV1GeneratorName:
+		generator = &kubectl.RoleGeneratorV1{
+			Name:          c.Name,
+			Verbs:         c.Verbs,
+			Resources:     c.Resources,
+			ResourceNames: c.ResourceNames,
+			APIGroups:     c.APIGroups,
+			Mapper:        mapper,
+		}
+	default:
+		return cmdutil.UsageError(cmd, fmt.Sprintf("Generator: %s not supported.", generatorName))
+	}
+
+	namespace, _, err := f.DefaultNamespace()
+	if err != nil {
+		return err
+	}
+
+	obj, err := generator.StructuredGenerate()
+	if err != nil {
+		return err
+	}
+
+	role := obj.(*rbac.Role)
+
+	// Transfer resource name from short form to long form.
+	resources := []string{}
+	for _, r := range role.Rules[0].Resources {
+		resource, err := mapper.ResourceFor(schema.GroupVersionResource{Resource: r})
```

Comment:
> xingzhou 13 hours ago Contributor
> Do you mean that for jobs.batch and hap.autoscaling, we need to hard code checking them and create two rules? Why these two need to be taken special care of?

Sorry, those aren't special, its just an example.  Because of the rule expansions, "get on (jobs,hpa) in (batch, autoscaling)" expands to 
 1. get on jobs.batch
 2. get on jobs.autoscaling
 3. get on hpa.batch
 4. get on hpa.autoscaling

That's an unexpected result.  You need to split the rule by API groups.